File: ruby/src/Structural/Facade/FacadeTeaCup.rb

# one of three classes the facade calls

class FacadeTeaCup

	def initialize()
		setTeaBagIsSteeped(false)
	end

	def setTeaBagIsSteeped(isTeaBagSteeped)
  		@teaBagIsSteeped = isTeaBagSteeped
	end

	def getTeaBagIsSteeped()
 		return @teaBagIsSteeped
	end

	def addFacadeTeaBag(facadeTeaBagIn)
 		@facadeTeaBag = facadeTeaBagIn
		return "the tea bag is in the tea cup"
	end

	def addFacadeWater(facadeWaterIn)
 		@facadeWater = facadeWaterIn
		return "the water is in the tea cup"
	end

	def steepTeaBag()
 		if @facadeTeaBag and @facadeWater and @facadeWater.getWaterIsBoiling()
 			setTeaBagIsSteeped(true)
			return "the tea is steeping in the cup"
		end
		setTeaBagIsSteeped(false)
		return "the tea is not steeping in the cup"
	end

	def to_s()
 		if getTeaBagIsSteeped()
 			return "A nice cuppa tea!" 
		end
		retval = ["A cup with "]
		if @facadeWater
 			if @facadeWater.getWaterIsBoiling()
 				retval.push("boiling water ")
			else
 				retval.push("cold water ")
			end
		else
 			retval.push("no water ")
		end

		if @facadeTeaBag
 			retval.push("and a tea bag")
		else
 			retval.push("and no tea bag")
		end
		return ''.join(retval)
	end

end

File: csharp/src/Structural/Facade/FacadeTeaCup.cs
namespace Structural.Facade {


//FacadeTeaCup.java - one of three classes the facade calls

public class FacadeTeaCup {  
   bool teaBagIsSteeped; 
   FacadeWater facadeWater;
   FacadeTeaBag facadeTeaBag;
    
   public FacadeTeaCup() {
       setTeaBagIsSteeped(false);
   }    
   
   public void setTeaBagIsSteeped(bool isTeaBagSteeped) {
       teaBagIsSteeped = isTeaBagSteeped;
   }
   public bool getTeaBagIsSteeped() {
       return teaBagIsSteeped;
   }
    
   public void addFacadeTeaBag(FacadeTeaBag facadeTeaBagIn) {
       facadeTeaBag = facadeTeaBagIn;
   }
   
   public void addFacadeWater(FacadeWater facadeWaterIn) {
       facadeWater = facadeWaterIn;
   }   
   
   public void steepTeaBag() {
       if ( (facadeTeaBag != null) && 
             ( (facadeWater != null) && 
               (facadeWater.getWaterIsBoiling()) )
          ) {
          setTeaBagIsSteeped(true);
       } else {
          setTeaBagIsSteeped(false);
       }           
   }
   
   override public string ToString() {
       if (this.getTeaBagIsSteeped()) {
           return ("A nice cuppa tea!");
       } else {
           string tempString = "A cup with ";
           if (facadeWater != null) {
                if (facadeWater.getWaterIsBoiling()) {
                    tempString = (tempString + "boiling water ");
                } else {
                    tempString = (tempString + "cold water ");
                }
           } else {
                tempString = (tempString + "no water ");
           }
 
           if (facadeTeaBag != null) {
                tempString = (tempString + "and a tea bag");
           } else {
                tempString = (tempString + "and no tea bag");
           } 
           return tempString;
       }
                        
   }
}

}
